dvc/fs/dvc.py

```diff
@@ -360,38 +360,47 @@ def isdvc(self, path, **kwargs) -> bool:
         except FileNotFoundError:
             return False
 
-    def ls(self, path, detail=True, dvc_only=False, **kwargs):
+    def ls(self, path, detail=True, dvc_only=False, **kwargs):  # noqa: C901
         key = self._get_key_from_relative(path)
         repo, dvc_fs, subkey = self._get_subrepo_info(key)
 
-        dvc_exists = False
         dvc_infos = {}
+        dvc_info = {}
         if dvc_fs:
             dvc_path = _get_dvc_path(dvc_fs, subkey)
-            with suppress(FileNotFoundError, NotADirectoryError):
-                for info in dvc_fs.ls(dvc_path, detail=True):
-                    dvc_infos[dvc_fs.name(info["name"])] = info
-                dvc_exists = True
+            with suppress(FileNotFoundError):
+                dvc_info = dvc_fs.info(dvc_path)
+                if dvc_info["type"] == "file":
+                    dvc_infos[""] = dvc_info
```

Comment:
Since `fs.ls()` is now expected to return result even if it's a file, we are keeping it in a special `""` (root) key so that we can return result for the file easily.

Later, we check if the type of the path matches in both `fs.info()` and `dvc_fs.info()`.
If they mismatch (eg: when a path `data` is tracked as a file in dvc, but is now a directory in the workspace or vice versa), we invalidate `dvc_fs.info()` and then return `fs.info()`.
This is the behaviour that we have in the tests.